cmd/kubelet/app/server.go

```diff
@@ -412,13 +430,218 @@ func run(s *options.KubeletServer, kcfg *KubeletConfig) (err error) {
+		return nil, fmt.Errorf("unable to find the name of the authInfo in current context %q", config.CurrentContext)
+	}
+
+	if _, ok := config.AuthInfos[ctx.AuthInfo]; !ok {
+		config.AuthInfos[ctx.AuthInfo] = clientcmdapi.NewAuthInfo()
+	}
+	return config.AuthInfos[ctx.AuthInfo], nil
+}
+
+// TODO(yifan): More detailed check on the cert / key content.
+// CheckTLSInfo returns true if the authInfo contains client certificate data for client key data.
+// Or if the client certificate or key file exists.
+func containsSufficientTLSInfo(authInfo *clientcmdapi.AuthInfo) bool {
+	// We use '||' so that we won't override existing data in case of wrong setup.
+	if len(authInfo.ClientCertificateData) > 0 || len(authInfo.ClientKeyData) > 0 {
+		return true
+	}
+
+	if crypto.FoundCertOrKey(authInfo.ClientCertificate, authInfo.ClientKey) {
+		return true
+	}
+
+	return false
+}
+
+// getCertFromAPIServer will:
+// (1) Create a restful client for doing the certificate signing request.
+// (2) Generate key pair and certificate signing request.
+//     The private key is stored to the given path.
+// (3) Send request to API server and watch for the issued certificate.
+// (4) Once (3) succeeds, dump the certificate to the given path.
+// On failure, the key and the certificate file will be cleaned up.
+// If the certfile or keyfile is empty, then it will use the default path, respectively:
+// /var/run/kubernetes/kubelet-client.srt, /var/run/kubernetes/kubelet-client.key.
+func getCertFromAPIServer(s *options.KubeletServer, certfile, keyfile string) (certPath, keyPath string, err error) {
+	certPath, keyPath = certfile, keyfile
+
+	// Cleanup keyfile and certfile.
+	defer func() {
+		if err != nil {
+			// Clean up cert and key.
+			if rmErr := os.Remove(certPath); rmErr != nil {
+				glog.Warningf("Failed to clean up TLS cert file %q: %v", certPath, rmErr)
+			}
+			if rmErr := os.Remove(keyPath); rmErr != nil {
+				glog.Warningf("Failed to clean up TLS private key file %q: %v", keyPath, rmErr)
+			}
+		}
+	}()
+
+	if certPath == "" {
+		certPath = defaultKubeletClientCertificateFile
+		if err := os.MkdirAll(filepath.Dir(certPath), 0644); err != nil {
+			return "", "", fmt.Errorf("unable to create dir for certificate file: %v", err)
+		}
+	}
+
+	if keyPath == "" {
+		keyPath = defaultKubeletClientKeyFile
+		if err := os.MkdirAll(filepath.Dir(keyPath), 0644); err != nil {
+			return "", "", fmt.Errorf("unable to create dir for key file: %v", err)
+		}
+	}
+
+	// (1).
+	clientConfig, err := kubeconfigClientConfig(s.BootstrapKubeconfig, s.APIServerList)
+	if err != nil {
+		return "", "", fmt.Errorf("unable to create client config: %v", err)
+	}
+	client, err := unversionedcertificates.NewForConfig(clientConfig)
+	if err != nil {
+		return "", "", fmt.Errorf("unable to create certificates signing request client: %v", err)
+	}
+
+	// (2).
+	hostname := nodeutil.GetHostname(s.HostnameOverride)
+
+	var ips []net.IP
+	nodeIP := net.ParseIP(s.NodeIP)
```

Comment:
node-ip is not required when using cloudprovider, and is in fact ignored when using cloudprovider.

I think it might be easier to get the Node object and build a cert based on those values, rather than trying to reimplement the somewhat complex logic here.